proposals/168/making_spark_visitors_extensible.md

```diff
@@ -0,0 +1,271 @@
+---
+Author: Michael Collado
+Created: December 14
+Issue: https://github.com/OpenLineage/OpenLineage/issues/168
+---
+
+# Purpose
+Users need to write custom extensions to the Spark OpenLineage library in order to support custom datasources, add
+custom job facets, and to augment events with application-specific logic. Currently, all of the logic for reporting
+OpenLineage events is contained in the `io.openlineage.spark.agent.lifecycle` package of the java library. This proposal
+allows for scanning the classpath to find custom implementations of event handlers that can publish OpenLineage 
+components based on application events.
+
+# Implementation
+
+Most of the code that publishes OpenLineage information from Spark applications uses Spark SQL's `LogicalPlan` to 
+locate and extract dataset and job information, such as the location of a dataset in GCS or the name and schema of a
+table in Hive or a relational database. Implementations subclass an abstract class called `QueryPlanVisitor`, which 
+is a `PartialFunction` that can convert specific implementations of `LogicalPlan` into `OpenLineage.Dataset`s. 
+However, some information necessary to construct a complete OpenLineage event is not found in the `LogicalPlan`. Simple
+examples include the number of records and bytes read and/or written during execution. That information is found in the
+`TaskMetrics` object bound to a `StageInfo` reported at the time of a `SparkListenerStageCompleted` event. Some 
+information, such as the `HadoopRDD` being processed is found in the `RDD` bound to a `Stage` (not a `StageInfo`).
+Thus, adding support for customer `QueryPlanVisitor`s will not allow full customization for users who are interested in
+publishing custom dataset or job/run facets. 
+
+To allow full customization, we can introduce the following interfaces
+```java
+/**
+ * Interface for custom facet builders. `isDefinedAt` determines the type of event supported by the 
+ */
+public interface CustomFacetBuilder<T, F> {
+
+    /**
+     * @param event
+     * @return true if this function is defined for the argument
+     */
+    boolean isDefinedAt(T event);
+
+    /**
+     * Handle the given event, writing any resulting facets to the {@link java.util.function.BiConsumer}, which accepts
+     * the name of the facet and the facet itself.
+     */
+    void accept(T event, BiConsumer<String, ? super F> consumer);
+
+}
+
+/**
+ * Factory for the builders that generate OpenLineage components and facets from Spark events.
+ */
+public interface OpenLineageEventListenerFactory {
```

Comment:
Based on https://github.com/OpenLineage/OpenLineage/pull/448, this is now `OpenLineageEventHandlerFactory`?

Reply:
Yeah, that's right. I have to go back to the proposal and update for changes necessitated by implementation